OverloadException on shutdownCluster

trying to shutdown my cluster but hazelcast cluster gives HazelcastOverloadException,
and will not do a graceful shutdown when requested.

calling cluster shutdown is sync call and could be considered 'high priority', 


https://hazelcast-l337.ci.cloudbees.com/view/stable/job/stable-cleanShutdown/2/console

/disk1/jenkins/workspace/stable-cleanShutdown/3.11-SNAPSHOT/2018_09_10-16_15_03/stable/async/async-bak/put-hd/big-val


```
fail HzMember1HZ clusterShutdown hzcmd.ops.ClusterShutDown threadId=0 com.hazelcast.core.HazelcastOverloadException: Failed to start invocation due to overload: Invocation{op=com.hazelcast.internal.cluster.impl.operations.LockClusterStateOp{serviceName='hz:core:clusterService', identityHash=682232950, partitionId=-1, replicaIndex=0, callId=0, invocationTime=1536586084033 (2018-09-10 13:28:04.033), waitTimeout=-1, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1536586084031, firstInvocationTime='2018-09-10 13:28:04.031', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.87]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
```

http://54.82.84.143/~jenkins/workspace/stable-cleanShutdown/3.11-SNAPSHOT/2018_09_10-16_15_03/stable/async/async-bak/put-hd/big-val


```
cat ./output/HZ/HzMember1HZ/exception.txt
com.hazelcast.core.HazelcastOverloadException: Failed to start invocation due to overload: Invocation{op=com.hazelcast.internal.cluster.impl.operations.LockClusterStateOp{serviceName='hz:core:clusterService', identityHash=682232950, partitionId=-1, replicaIndex=0, callId=0, invocationTime=1536586084033 (2018-09-10 13:28:04.033), waitTimeout=-1, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1536586084031, firstInvocationTime='2018-09-10 13:28:04.031', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.87]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationRegistry.register(InvocationRegistry.java:117)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:559)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:530)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:220)
	at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.invokeOnTarget(OperationServiceImpl.java:327)
	at com.hazelcast.internal.cluster.impl.ClusterStateManager.lockClusterStateOnAllMembers(ClusterStateManager.java:413)
	at com.hazelcast.internal.cluster.impl.ClusterStateManager.changeClusterState(ClusterStateManager.java:367)
	at com.hazelcast.internal.cluster.impl.ClusterStateManager.changeClusterState(ClusterStateManager.java:345)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.changeClusterState(ClusterServiceImpl.java:854)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.shutdownCluster(ClusterServiceImpl.java:933)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.shutdown(ClusterServiceImpl.java:923)
	at hzcmd.ops.ClusterShutDown.timeStep(ClusterShutDown.java:8)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:42)
	at remote.bench.BenchThread.call(BenchThread.java:15)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.core.HazelcastOverloadException: Timed out trying to acquire another call ID. maxConcurrentInvocations = 9, backoffTimeout = 60000 msecs, elapsed:60169 msecs
	at com.hazelcast.spi.impl.sequence.CallIdSequenceWithBackpressure.handleNoSpaceLeft(CallIdSequenceWithBackpressure.java:61)
	at com.hazelcast.spi.impl.sequence.AbstractCallIdSequence.next(AbstractCallIdSequence.java:56)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationRegistry.register(InvocationRegistry.java:115)
	... 21 more
com.hazelcast.core.HazelcastOverloadException: Failed to start invocation due to overload: Invocation{op=com.hazelcast.internal.cluster.impl.operations.LockClusterStateOp{serviceName='hz:core:clusterService', identityHash=682232950, partitionId=-1, replicaIndex=0, callId=0, invocationTime=1536586084033 (2018-09-10 13:28:04.033), waitTimeout=-1, callTimeout=60000}, tryCount=250, tryPauseMillis=500, invokeCount=1, callTimeoutMillis=60000, firstInvocationTimeMs=1536586084031, firstInvocationTime='2018-09-10 13:28:04.031', lastHeartbeatMillis=0, lastHeartbeatTime='1970-01-01 00:00:00.000', target=[10.0.0.87]:5701, pendingResponse={VOID}, backupsAcksExpected=0, backupsAcksReceived=0, connection=null}
	at com.hazelcast.spi.impl.operationservice.impl.InvocationRegistry.register(InvocationRegistry.java:117)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:559)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:530)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:220)
	at com.hazelcast.spi.impl.operationservice.impl.OperationServiceImpl.invokeOnTarget(OperationServiceImpl.java:327)
	at com.hazelcast.internal.cluster.impl.ClusterStateManager.lockClusterStateOnAllMembers(ClusterStateManager.java:413)
	at com.hazelcast.internal.cluster.impl.ClusterStateManager.changeClusterState(ClusterStateManager.java:367)
	at com.hazelcast.internal.cluster.impl.ClusterStateManager.changeClusterState(ClusterStateManager.java:345)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.changeClusterState(ClusterServiceImpl.java:854)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.shutdownCluster(ClusterServiceImpl.java:933)
	at com.hazelcast.internal.cluster.impl.ClusterServiceImpl.shutdown(ClusterServiceImpl.java:923)
	at hzcmd.ops.ClusterShutDown.timeStep(ClusterShutDown.java:8)
	at remote.bench.marker.MetricsMarker.flatOut(MetricsMarker.java:53)
	at remote.bench.marker.MetricsMarker.bench(MetricsMarker.java:40)
	at remote.bench.BenchThread.call(BenchThread.java:42)
	at remote.bench.BenchThread.call(BenchThread.java:15)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.hazelcast.core.HazelcastOverloadException: Timed out trying to acquire another call ID. maxConcurrentInvocations = 9, backoffTimeout = 60000 msecs, elapsed:60169 msecs
	at com.hazelcast.spi.impl.sequence.CallIdSequenceWithBackpressure.handleNoSpaceLeft(CallIdSequenceWithBackpressure.java:61)
	at com.hazelcast.spi.impl.sequence.AbstractCallIdSequence.next(AbstractCallIdSequence.java:56)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationRegistry.register(InvocationRegistry.java:115)
	... 21 more
```

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/CommitClusterStateOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/CommitClusterStateOp.java
@@ -27,6 +27,7 @@ import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.spi.ExceptionAction;
 import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.UrgentSystemOperation;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 import com.hazelcast.spi.impl.AllowedDuringPassiveState;
 import com.hazelcast.transaction.TransactionException;
@@ -35,7 +36,8 @@ import java.io.IOException;
 
 import static java.lang.String.format;
 
-public class CommitClusterStateOp extends Operation implements AllowedDuringPassiveState, IdentifiedDataSerializable {
+public class CommitClusterStateOp extends Operation implements AllowedDuringPassiveState, UrgentSystemOperation,
+        IdentifiedDataSerializable {
 
     private ClusterStateChange stateChange;
     private Address initiator;

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/LockClusterStateOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/LockClusterStateOp.java
@@ -28,13 +28,15 @@ import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.spi.ExceptionAction;
 import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.UrgentSystemOperation;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 import com.hazelcast.spi.impl.AllowedDuringPassiveState;
 import com.hazelcast.transaction.TransactionException;
 
 import java.io.IOException;
 
-public class LockClusterStateOp  extends Operation implements AllowedDuringPassiveState, IdentifiedDataSerializable {
+public class LockClusterStateOp  extends Operation implements AllowedDuringPassiveState, UrgentSystemOperation,
+        IdentifiedDataSerializable {
 
     private ClusterStateChange stateChange;
     private Address initiator;

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/RollbackClusterStateOp.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/RollbackClusterStateOp.java
@@ -26,12 +26,14 @@ import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
 import com.hazelcast.spi.ExceptionAction;
 import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.UrgentSystemOperation;
 import com.hazelcast.spi.exception.TargetNotMemberException;
 import com.hazelcast.spi.impl.AllowedDuringPassiveState;
 
 import java.io.IOException;
 
-public class RollbackClusterStateOp extends Operation implements AllowedDuringPassiveState, IdentifiedDataSerializable {
+public class RollbackClusterStateOp extends Operation implements AllowedDuringPassiveState, UrgentSystemOperation,
+        IdentifiedDataSerializable {
 
     private Address initiator;
     private String txnId;

--- a/hazelcast/src/test/java/com/hazelcast/cluster/ClusterShutdownTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cluster/ClusterShutdownTest.java
@@ -16,9 +16,18 @@
 
 package com.hazelcast.cluster;
 
+import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.HazelcastOverloadException;
 import com.hazelcast.instance.Node;
 import com.hazelcast.instance.NodeState;
+import com.hazelcast.nio.Address;
+import com.hazelcast.spi.AbstractWaitNotifyKey;
+import com.hazelcast.spi.BlockingOperation;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.WaitNotifyKey;
+import com.hazelcast.spi.impl.operationservice.InternalOperationService;
+import com.hazelcast.spi.properties.GroupProperty;
 import com.hazelcast.test.AssertTask;
 import com.hazelcast.test.HazelcastParallelClassRunner;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -29,9 +38,14 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
+import java.util.concurrent.Callable;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeoutException;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
 
 @RunWith(HazelcastParallelClassRunner.class)
 @Category({QuickTest.class, ParallelTest.class})
@@ -67,6 +81,40 @@ public class ClusterShutdownTest extends HazelcastTestSupport {
         HazelcastInstance instance = testClusterShutdownWithSingleMember(ClusterState.ACTIVE);
 
         instance.getLifecycleService().shutdown();
+    }
+
+    @Test
+    public void clusterShutdown_shouldNotBeRejected_byBackpressure() throws Exception {
+        Config config = new Config();
+        config.setProperty(GroupProperty.PARTITION_COUNT.toString(), "1");
+        config.setProperty(GroupProperty.BACKPRESSURE_ENABLED.toString(), "true");
+        config.setProperty(GroupProperty.BACKPRESSURE_BACKOFF_TIMEOUT_MILLIS.toString(), "100");
+        config.setProperty(GroupProperty.BACKPRESSURE_MAX_CONCURRENT_INVOCATIONS_PER_PARTITION.toString(), "3");
+
+        HazelcastInstance hz = createHazelcastInstance(config);
+        final InternalOperationService operationService = getOperationService(hz);
+        final Address address = getAddress(hz);
+
+        for (int i = 0; i < 10; i++) {
+            Future<Object> future = spawn(new Callable<Object>() {
+                @Override
+                public Object call() {
+                    operationService.invokeOnTarget(null, new AlwaysBlockingOperation(), address);
+                    return null;
+                }
+            });
+            try {
+                future.get();
+            } catch (ExecutionException e) {
+                assertInstanceOf(HazelcastOverloadException.class, e.getCause());
+            }
+        }
+
+        Node node = getNode(hz);
+        hz.getCluster().shutdown();
+
+        assertFalse(hz.getLifecycleService().isRunning());
+        assertEquals(NodeState.SHUT_DOWN, node.getState());
     }
 
     private HazelcastInstance testClusterShutdownWithSingleMember(ClusterState clusterState) {
@@ -122,11 +170,38 @@ public class ClusterShutdownTest extends HazelcastTestSupport {
         for (final Node node : nodes) {
             assertTrueEventually(new AssertTask() {
                 @Override
-                public void run()
-                        throws Exception {
+                public void run() {
                     assertEquals(NodeState.SHUT_DOWN, node.getState());
                 }
             });
+        }
+    }
+
+    private static class AlwaysBlockingOperation extends Operation implements BlockingOperation {
+
+        @Override
+        public void run() throws Exception {
+        }
+
+        @Override
+        public WaitNotifyKey getWaitKey() {
+            return new AbstractWaitNotifyKey(getServiceName(), "test") {
+            };
+        }
+
+        @Override
+        public boolean shouldWait() {
+            return true;
+        }
+
+        @Override
+        public void onWaitExpire() {
+            sendResponse(new TimeoutException());
+        }
+
+        @Override
+        public String getServiceName() {
+            return "AlwaysBlockingOperationService";
         }
     }
 }

